phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,547 @@
+
+            if (inputTables == null) {
+                inputTables = new String(Files.readAllBytes(Paths.get(inputFile)));
+            }
+            if (inputTables == null) {
+                LOGGER.severe("Tables' list is not available; use -tb or -f option");
+            }
+            LOGGER.info("list of tables passed: " + inputTables);
+
+            if (operation.equalsIgnoreCase(UPGRADE_OP)) {
+                upgrade = true;
+            } else if (operation.equalsIgnoreCase(ROLLBACK_OP)) {
+                upgrade = false;
+            } else {
+                throw new IllegalStateException("Invalid option provided for "
+                        + OPERATION_OPTION.getOpt() + " expected values: {upgrade, rollback}");
+            }
+            if (dryRun) {
+                LOGGER.info("This is the beginning of the tool with dry run.");
+            }
+        } catch (IOException e) {
+            LOGGER.severe("Something went wrong "+e);
+            System.exit(-1);
+        }
+    }
+
+    @VisibleForTesting
+    public void executeTool() {
+        try {
+            Configuration conf = HBaseConfiguration.addHbaseResources(getConf());
+            Connection conn = ConnectionUtil.getInputConnection(conf);
+
+            ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class)
+                    .getQueryServices();
+            extractTablesAndIndexes(conn.unwrap(PhoenixConnection.class));
+
+            executeTool(conn, queryServices, conf);
+        } catch (SQLException e) {
+            LOGGER.severe("Something went wrong in executing tool "+ e);
+            System.exit(-1);
+        }
+    }
+
+    private void executeTool(Connection conn, ConnectionQueryServices queryServices,
+            Configuration conf) {
+
+        LOGGER.info("Executing "+ operation);
+
+        for (Map.Entry<String, HashSet<String>> entry: tablesAndIndexes.entrySet()) {
+            String dataTableFullName = entry.getKey();
+            HashSet<String> indexes = entry.getValue();
+
+            try (Admin admin = queryServices.getAdmin()) {
+
+                PTable dataTable = PhoenixRuntime.getTableNoCache(conn, dataTableFullName);
+                LOGGER.fine("Executing "+operation+" for " + dataTableFullName);
+
+                boolean mutable = !(dataTable.isImmutableRows());
+                if (!mutable) {
+                    LOGGER.fine("Data table is immutable, waiting for "+GLOBAL_INDEX_CHECKER_ENABLED_MAP_EXPIRATION_MIN + 1+" minutes for client cache to expire");
+                    if(!test) {
+                        Thread.sleep(
+                                (GLOBAL_INDEX_CHECKER_ENABLED_MAP_EXPIRATION_MIN + 1) * 60 * 1000);
+                    }
+                }
+                if (upgrade) {
+                    disableTable(admin, dataTableFullName, indexes);
+                    modifyIndexTable(admin, indexes);
+                    if (mutable) {
```

Comment:
We should modify the coprocessors on the data table for immutable tables too as eventually we will remove the Indexer coprocessor from the code base.